### PR TITLE
Add functional tests

### PR DESCRIPTION
**Pull request by sbs2001**
Functional Tests are implemented as follows:

1. A mock LAPI is implemented using flask. The bouncer is pointed towards it. 
2. Decisions are inserted/deleted in this LAPI and then the resultant state is asserted using `nft ..` commands.

### CODE REVIEW

**Comment by mmetc** [comment]
Can you trim down the .gitignore? It's huge.. and keep the full configuration name, for consistency
Maybe a Make target to create the virtualenv and run the tests

Ideally, it should be possible to run tests as non-root (polkit?) but we can see that later

File: tests/nftables/test_nftables.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+import os
+import sys
+from time import sleep
+from typing import List
+from ipaddress import ip_address
+
+from tests.nftables import mock_lapi
+
+SCRIPT_DIR = os.path.dirname(os.path.realpath(__file__))
+os.chdir(SCRIPT_DIR)
```
Comment (mmetc) [inline tests/nftables/test_nftables.py:14]: These global-level function calls are run when you import a module, possibly causing side effects. Better to change directory in the explicit control flow (i.e. main() or whatever is the entry point). May not be a problem with two files, but when you have multiple os.chdir() of this kind, import order can trigger bugs

File: tests/nftables/test_nftables.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        set_elements = self.get_set_elements("ip", "crowdsec", "crowdsec-blacklists")
+        self.assertEqual(len(set_elements), total_decisions - duplicate_decisions)
+        assert set([i["value"] for i in decisions]) == set_elements
+        assert "0.0.0.0" in set_elements
+
+        mock_lapi.ds.delete_decisions_by_ip("0.0.0.0")
+        sleep(1)
+
+        set_elements = self.get_set_elements("ip", "crowdsec", "crowdsec-blacklists")
+        assert set([i["value"] for i in decisions if i["value"] != "0.0.0.0"]) == set_elements
```
Comment (mmetc) [inline tests/nftables/test_nftables.py:147]: python has set comprehension: set(i["value"] ... "0.0.0.0"), no need to create a list. Or literal: {i["value"].... "0.0.0.0"}

File: tests/nftables/test_nftables.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+            print(cmd, "exited with non-zero code with following logs")
+            print(p.stdout.read().decode())
+            sys.exit(1)
+        else:
+            return p.stdout.read().decode()
+
+
+def generate_n_decisions(n: int, action="ban", dup_count=0, ipv4=True):
+    if dup_count >= n:
+        print("generate_n_decisions got dup_count>=n")
```
Comment (mmetc) [inline tests/nftables/test_nftables.py:29]: print + exit(1) is a sign that maybe the output is intended for stderr. Ask yourself: is it useful in a pipe, in a script? Or is it a message/warning/error for a human? In the latter case, return the message with an exception. If you don't want or need to create your exceptions, I think "raise SystemExit(message)" is safe to use and sets $? to 1.

**Comment by mmetc** [comment]
to sum up what we said

1. use the default values for mapping.get() to make intentions clear: `if "set" not in node or "elem" not in node["set"]`  -> `'elem' not in a.get('set', {})`
2. use exceptions for flow control instead of too many if/else (which is a good pattern in other languages, less so in python) - https://devblogs.microsoft.com/python/idiomatic-python-eafp-versus-lbyl/
3. separate the methods where you **use** objects from the methods that **change** objects (not sure if there's still any in this code)
4. staticmethods are almost always better as a stand-alone function, they do the same thing and are more visible/readable
5. avoid enumerate() if you don't really need the index
6. the imports on the top should have this order:  standard library, empty line, third party packages (i.e. dependencies), empty line, then you own packages

**Review by buixor — APPROVED** [review]
lgtm

**Pull request merged**